examples/ini.rs

```diff
@@ -0,0 +1,292 @@
+    name: &'a str,
+    properties: Vec<Pair<'a>>,
+}
+
+fn read_ini<'i, E>(r: &mut Reader<'i, E>) -> Result<Document<'i>, E>
+where
+    E: Error<'i>,
+{
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    if r.at_end() {
+        return Ok(Document::default());
+    }
+    let (globals, sections) = match r.peek_u8()? {
+        b'[' => (vec![], read_sections(r)?),
+        _ => (
+            read_zero_or_more_properties_until_section(r)?,
+            read_sections(r)?,
+        ),
+    };
+    Ok(Document { globals, sections })
+}
+
+fn read_sections<'i, E>(r: &mut Reader<'i, E>) -> Result<Vec<Section<'i>>, E>
+where
+    E: Error<'i>,
+{
+    let mut sections = Vec::new();
+    while !r.at_end() {
+        sections.push(read_section(r)?);
+    }
+    Ok(sections)
+}
+
+fn read_zero_or_more_properties_until_section<'i, E>(
+    r: &mut Reader<'i, E>,
+) -> Result<Vec<Pair<'i>>, E>
+where
+    E: Error<'i>,
+{
+    let mut out = Vec::new();
+    fn is_bare_text(c: u8) -> bool {
+        !c.is_ascii_whitespace() && c != b'=' && c != b'\n' && c != b'['
+    }
+
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    while !(r.at_end() || matches!(r.peek_u8(), Ok(b'['))) {
+        r.context("property", |r| {
+            skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+            let name = r.context("name", |r| {
+                r.take_while(is_bare_text).to_dangerous_non_empty_str()
+            })?;
+            skip_whitespace_or_comment(r, ConsumeTo::EndOfLine);
+
+            r.consume_u8(b'=')?;
+
+            skip_whitespace_or_comment(r, ConsumeTo::EndOfLine);
+            let value = r.context("value", |r| {
+                r.take_while(|c| c != b';' && c != b'\n' && c != b'=' && c != b'[')
+                    .to_dangerous_non_empty_str()
+                    .map(str::trim)
+            })?;
+            skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+            out.push(Pair { name, value });
+            Ok(())
+        })?;
+    }
+    Ok(out)
+}
+
+fn read_section<'i, E>(r: &mut Reader<'i, E>) -> Result<Section<'i>, E>
+where
+    E: Error<'i>,
+{
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    r.consume_u8(b'[')?;
+    let name = r
+        .context("section name", |r| {
+            r.take_while(|c| c != b']' && c != b'\n')
+                .to_dangerous_non_empty_str()
+        })?
+        .trim();
+    r.consume_u8(b']')?;
+
+    r.try_expect::<_, ()>("newline after section", |r| {
+        let past_section = r.take_while(|c| c.is_ascii_whitespace());
+        if past_section.as_dangerous().contains(&b'\n') {
+            Ok(Some(()))
+        } else {
+            Ok(None)
+        }
+    })?;
+    let properties = read_zero_or_more_properties_until_section(r)?;
+
+    Ok(Section { name, properties })
+}
+
+enum ConsumeTo {
+    NextToken,
+    EndOfLine,
+}
+
+fn skip_whitespace_or_comment<'i, E>(r: &mut Reader<'i, E>, to_where: ConsumeTo)
+where
+    E: Error<'i>,
```

Comment:
The generic constraint for `Error<'i>` can be removed here :)

Reply:
Oh wow, much better! I guess it was required at some point, until it wasn't anymore.

Applied!